include/zephyr/drivers/retained_mem.h

```diff
@@ -0,0 +1,198 @@
+ * See retained_mem_read() for argument description.
+ */
+typedef int (*retained_mem_read_api)(const struct device *dev, off_t offset, uint8_t *buffer,
+				     size_t size);
+
+/**
+ * @typedef	retained_mem_write_api
+ * @brief	Callback API to write to retained memory area.
+ * See retained_mem_write() for argument description.
+ */
+typedef int (*retained_mem_write_api)(const struct device *dev, off_t offset,
+				      const uint8_t *buffer, size_t size);
+
+/**
+ * @typedef	retained_mem_clear_api
+ * @brief	Callback API to clear retained memory area (reset all data to 0x00).
+ * See retained_mem_clear() for argument description.
+ */
+typedef int (*retained_mem_clear_api)(const struct device *dev);
+
+/**
+ * @brief Retained memory driver API
+ * API which can be used by a device to store data in a retained memory area. Retained memory is
+ * memory that is retained while the device is powered but is lost when power to the device is
+ * lost (note that low power modes in some devices may clear the data also). This may be in a
+ * non-initialised RAM region, or in specific registers, but is not reset when a different
+ * application begins execution or the device is rebooted (without power loss). It must support
+ * byte-level reading and writing without a need to erase data before writing.
+ *
+ * Note that drivers must implement all functions, none of the functions are optional.
+ */
+struct retained_mem_driver_api {
+	retained_mem_size_api size;
+	retained_mem_read_api read;
+	retained_mem_write_api write;
+	retained_mem_clear_api clear;
+};
+
+/**
+ * @brief		Returns the size of the retained memory area.
+ *
+ * @param dev		Retained memory device to use.
+ *
+ * @retval		Positive value indicating size in bytes on success, else negative errno
+ *			code.
+ */
+__syscall ssize_t retained_mem_size(const struct device *dev);
+
+static inline ssize_t z_impl_retained_mem_size(const struct device *dev)
+{
+	struct retained_mem_driver_api *api = (struct retained_mem_driver_api *)dev->api;
+
+	return api->size(dev);
+}
+
+/**
+ * @brief		Reads data from the Retained memory area.
+ *
+ * @param dev		Retained memory device to use.
+ * @param offset	Offset to read data from.
+ * @param buffer	Buffer to store read data in.
+ * @param size		Size of data to read.
+ *
+ * @retval		0 on success else negative errno code.
+ */
+__syscall int retained_mem_read(const struct device *dev, off_t offset, uint8_t *buffer,
+				size_t size);
+
+static inline int z_impl_retained_mem_read(const struct device *dev, off_t offset,
+					   uint8_t *buffer, size_t size)
+{
+	struct retained_mem_driver_api *api = (struct retained_mem_driver_api *)dev->api;
+	size_t end;
+	bool overflow;
+
+	/* Validate user-supplied parameters */
+	if (size == 0) {
+		return 0;
+	} else if (offset < 0) {
+		return -EINVAL;
+	}
+
+	overflow = size_add_overflow((size_t)offset, size, &end);
+
+	if (overflow || end > api->size(dev)) {
+		return -EINVAL;
+	}
```

Comment:
This part requires more time to figure out than probably needed.
Why not just follow the normal way (proposed by @Laczen, thanks):
```
 if ((dev->size < len) || ((dev->size(dev) - len) < (size_t)offset)) {
                return -EINVAL;
        }
```

I do not even think that usage of `size_add_overflow`, IMHO this thing should only by used on things that can be directly accessed by indexing address space.
We have to use cast to even use this thing on offset, which may not be save.
Consider a case when `sizeof(off_t) > sizeof(size_t)`: this is explicit cast, so no warning will be reported that `off_t` type variable gets cut down to `sizeof(size_t)`. This  means that in scenarios where `(size_t)(offset) < size`, call to `size_add_overflow` will report no overflow and `end` will be in range of `api->size(dev)`, which means that `api->read(...)` will be called with `offset` that is out of range.

Reply:
This is how they are set in the arm zephyr sdk:
```
typedef long _off_t;
typedef unsigned long __size_t;
```
So off_t can never be larger than size_t, only smaller, and the above check exits out if the provided offset is less than 0.

Reply:
The `off_t` is not defined within C standard, so far, and comes from POSIX definition. `size_t` is architecture dependent because it is supposed to be able to give size of largest memory object that can be addressed.
There is no dependency between `off_t` and `size_t` because they come from different worlds.
(and my proposed code above has risk of truncation error too, due to size_t cast).

I seen no benefit of using `size_and_overflow` when:
 1) `size_and_overflow` using code is just ugly now and looks jagged;
 2) it is actually half working because it will work as long as `off_t` <= `size_t` and the hides potential problem, as used `offset` is not actually the one that is used for overflow check.